src/relax/transform/lift_transform_params.cc

```diff
@@ -37,405 +38,467 @@
- *
- * When a binding cannot be lifted, all the variables that 1) it depends on, and 2) have been
- * lifted, will be marked as the boundary variable and will be in the output of the lifted function.
- */
-class LiftTransformParamsPlanner : public ExprVisitor {
+class LiftableBindingCollector : ExprVisitor {
  public:
-  LiftTransformParamsInfoPlan Plan(const Function& function, int num_inputs) {
-    for (int i = 0; i < static_cast<int>(function->params.size()); ++i) {
-      if (i < num_inputs) {
-        builder_.UpdateBasedOnRuntimeInput(function->params[i]);
-      } else {
-        builder_.AddInput(function->params[i]);
-        if (function->params[i]->struct_info_.defined()) {
-          Array<tir::Var> symbolic_vars = DefinableTIRVarsInStructInfo(
-              Downcast<StructInfo>(function->params[i]->struct_info_.value()));
-          for (const auto& var : symbolic_vars) {
-            param_symbolic_vars_.insert(var);
-          }
-        }
-      }
+  static CollectInfo Collect(const Function& func) {
+    LiftableBindingCollector visitor;
+    visitor(func);
+    visitor.info_.orig_func = func;
+    return visitor.info_;
+  }
+
+ private:
+  void VisitExpr_(const FunctionNode* func) override {
+    size_t num_runtime_params = func->params.size();
+    if (auto opt = func->attrs.GetAttr<Integer>(attr::kNumInput)) {
+      num_runtime_params = opt.value()->value;
     }
-    VisitExpr(function->body);
 
-    const auto& [f_transform_params, output_to_index] = builder_.Build();
-    return {f_transform_params, output_to_index, std::move(builder_.lifted_binding_lookup_)};
+    info_.num_runtime_params = num_runtime_params;
+
+    for (size_t i = num_runtime_params; i < func->params.size(); i++) {
+      liftable_vars_.insert(func->params[i]);
+      info_.requires_compile_time_param.insert(func->params[i]);
+      for (const auto& tir_var : DefinableTIRVarsInStructInfo(GetStructInfo(func->params[i]))) {
+        liftable_vars_.insert(tir_var);
+      }
+    }
+    ExprVisitor::VisitExpr_(func);
   }
 
- private:
   void VisitBindingBlock_(const DataflowBlockNode* block) final {
+    bool cache = is_in_dataflow_block_;
     is_in_dataflow_block_ = true;
     ExprVisitor::VisitBindingBlock_(block);
-    is_in_dataflow_block_ = false;
+    is_in_dataflow_block_ = cache;
+  }
+
+  void VisitBinding(const Binding& binding) override {
+    auto bound_value = GetBoundValue(binding);
+
+    if (CanLiftBinding(binding)) {
+      info_.computable_at_compile_time.push_back(binding);
+      liftable_vars_.insert(binding->var);
+
+      auto upstream_vars = FreeVars(bound_value);
+      bool depends_on_compile_time_param = std::any_of(
+          upstream_vars.begin(), upstream_vars.end(),
+          [&](const Var& var) -> bool { return info_.requires_compile_time_param.count(var); });
+      if (depends_on_compile_time_param) {
+        info_.requires_compile_time_param.insert(binding->var);
+      }
```

Comment:
Is this the logic that is meant to handle zero-param operators? I had a really hard time figuring out what the change was from #16594 and this was the only major area that came up in the diff.

Reply:
This section along with [this line](https://github.com/apache/tvm/pull/16595/commits/a8ca845a883cfa7440941cba5ec35c11e2294bc8#diff-dbc6646d7cc42ae403e97adb88f092c13fd91bc928595b7f44d54a3f218d714bR124), yes.  This section collects information to determine if this parameter depends, directly or indirectly, on one of the model weights.  The linked line uses the collected information to determine if the `transform_params` should output a variable.

I've updated the comment here to indicate why the additional information is being collected.